### PR TITLE
Added the `>>` syntax for specifying output indices.

### LINKED ISSUE #6: Explicit specification of tensor expression output indices
The `einsum` API of common numerical linear algebra libraries such as NumPy, PyTorch, and JAX all support the extended Einstein notation format, which allows an explicit expression of the left-hand side indices in the specs string. This explicit specification is the part after the `->` symbol:
```
# implicit lhs, inferred to be ik
ij,jk

# explicit lhs
ij,jk->ik

# explicit lhs, result transposed
ij,jk->ki

# explicit lhs, disables summation over j
ij,jk->ijk
```
As such, the specification of left-hand side indices has the effect of transposition and explicitly enabling/disabling the reduction operation over certain indices.

Adding this functionality would entail semantics that may look like:
```
# A, B, C, D: tensors
# i, j, k: indices
C[k, i] = A[i, j] * B[j, k]
D[i, k, j] = A[i, j] * B[j, k]
```

### PR DESCRIPTION
**Pull request by yhtang**
Summary:
- Added the `>>` operator for the explicit specification of output tensor indices.
- A shorthand syntax, which introduces the `~` modifier to individual indices, will be implemented in #17.
- Replaces #16.
- Closes #6.

Example:
```
A = ff.tensor('A', 2, 3)
B = ff.tensor('B', 3, 4)
i, j, k = ff.indices('i, j, k')
tsrex = A[i, j] * B[j, k] >> (k, j)
```
give rise to the following tensor expression
```
 >>
 ├── *
 │   ├── A[i,j]
 │   │   ├── A
 │   │   ╰── i,j
 │   │       ├── i
 │   │       ╰── j
 │   ╰── B[j,k]
 │       ├── B
 │       ╰── j,k
 │           ├── j
 │           ╰── k
 ╰── k,j
     ├── k
     ╰── j
```
which in addition to carrying out the inner product, also transposes the final result.

This is currently a **work-in-progress**: a forward-backward interpreter is needed to implement index analysis for the new syntax.

### CODE REVIEW

File: funfact/lang/_tensor.py [hunk 1]
```diff
@@ -17,16 +17,20 @@ def symbol(self):
...
-                'underscore and a numeric subscript. '
-                'Examples: i, j, k_0, lhs, etc.'
-            )
-        self._letter, self._number = m.groups()
-        if self._letter == 'Anonymous':
-            self._letter = r'\varphi'
+        try:
+            self._letter, self._number = m.groups()
+        except AttributeError:
+            m = re.fullmatch(r'__(\d+)', string)
```
Comment (campsd) [inline funfact/lang/_tensor.py:23]: Why do you check for symbols with double underscores? Is this related to #4 ?

**Review by campsd — COMMENTED** [review]
In the `_index_propagation` interpreter the binary nodes should no longer just return the symmetric difference but use the information from the `let` node (if present).

File: funfact/lang/_tensor.py (hunk 1 above)
Comment (yhtang) [reply funfact/lang/_tensor.py:23]: Right. This is still a temporary solution. The plan is to allow truly anonymous indices, and then only assign readable letters to them when rendering in latex/ascii.

**Comment by yhtang** [comment]
> In the `_index_propagation` interpreter the binary nodes should no longer just return the symmetric difference but use the information from the `let` node (if present).

Yes. I'm working on an update which actually removes the `let` primitive, merged all the binary primitives as a single `ein` primitive, and make the output indices a builtin field of `ein`. This should eliminate the need for a backward interpreter when doing index/einspec analysis, and also allow future generalization of the binary operators to include things such as the minplus semiring.

File: funfact/lang/interpreter/_einop.py [hunk 2]
```diff
@@ -67,7 +79,11 @@ def _einop(spec: str, lhs, rhs, op):
     con_ax = tuple(con_ax)
 
     # compute the contraction in alphabetical order
-    result = np.sum(op(lhs[dim_lhs], rhs[dim_rhs]), axis=con_ax)
+    op_redu = getattr(DummyBackend, reduction)
+    op_pair = getattr(DummyBackend, pairwise)
+
+    print('op_redu', op_redu)
```
Comment (campsd) [inline funfact/lang/interpreter/_einop.py:85]: Should this be removed?

File: funfact/lang/_semiring.py [hunk 3]
```diff
@@ -0,0 +1,10 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+from ._ast import Primitives as P
+from ._tsrex import TsrEx, EinopEx
+
+
+def minplus(lhs: TsrEx, rhs: TsrEx):
```
Comment (campsd) [inline funfact/lang/_semiring.py:7]: This errors for me: 

`tsrex = minplus(A[i, j], B[j,k])` returns
`AttributeError: 'TsrEx' object has no attribute '_as_node'`

File: funfact/lang/_semiring.py [hunk 4]
```diff
@@ -0,0 +1,10 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+from ._ast import Primitives as P
+from ._tsrex import TsrEx, EinopEx
+
```
Comment (campsd) [inline funfact/lang/_semiring.py:5]: We can add support for log-spaced Einsum contractions.
We need  to add a reduction of the form shown below to the `DummyBackend` in `_einop.py` and create
an `EinopEx` for this.

```
def logspaceSum(data, axis=None):
    return np.log(np.sum(np.exp(data), axis=axis))
```

File: funfact/lang/_semiring.py (hunk 3 above)
Comment (campsd) [reply funfact/lang/_semiring.py:7]: The same error happens with the math operations in `_math.py` for which the `_as_node` function is used.
